cpp/src/arrow/array/array_nested.cc

```diff
@@ -836,7 +836,7 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
     return Status::NotImplemented("Null bitmap with offsets slice not supported.");
   }
 
-  if (offsets->null_count() > 0) {
+  if (offsets->data()->MayHaveNulls()) {
```

Comment:
There is a similar condition above `keys->null_count() != 0` that should be `keys->MayHaveNulls()` as well. I think you've changed this in the other PR (?) so a rebase here would do?